### PR TITLE
Remove many `CompatibilitySwitch<T>` properties

### PR DESCRIPTION
**Pull request by dougbu**
- #4866 2 of 3
- will leave one `IEnumerable<ICompatibilitySwitch>` implementation: `MvcOptions`
- will leave one `ConfigureCompatibilityOptions<MvcOptions>` subclass: `MvcOptionsConfigureCompatibilityOptions`
- a few options remain as regular properties:
  - `ApiBehaviorOptions.SuppressMapClientErrors` (default `false`)
  - `MvcOptions.EnableEndpointRouting` (default `true`)
  - `MvcOptions.MaxValidationDepth` (default `32`)
  - `MvcJsonOptions.AllowInputFormatterExceptionMessages` (default `true`)
- remains possible to disable many features
  - e.g. replace `IModelBinderProvider`s with ones that pass `allowValidatingTopLevelNodes: false`

### CODE REVIEW

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/InferParameterBindingInfoConvention.cs [hunk 1]
```diff
@@ -118,13 +116,7 @@ private bool ParameterExistsInAnyRoute(ActionModel action, string parameterName)
         private bool IsComplexTypeParameter(ParameterModel parameter)
         {
             // No need for information from attributes on the parameter. Just use its type.
-            var metadata = _modelMetadataProvider
-                .GetMetadataForType(parameter.ParameterInfo.ParameterType);
-
-            if (AllowInferringBindingSourceForCollectionTypesAsFromQuery && metadata.IsCollectionType)
```
Comment (dougbu) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/InferParameterBindingInfoConvention.cs]: `AllowInferringBindingSourceForCollectionTypesAsFromQuery` was one of the few properties that was `false` with recent `CompatibilityVersion`s.

**Comment by JamesNK** [comment]
> * MvcOptions.EnableEndpointRouting (default true)

I'm not super familiar with compat but this point is fine 👍

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs [hunk 2]
```diff
@@ -136,38 +134,30 @@ private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFil
             var effectivePolicy = await ComputePolicyAsync();
             var canCache = PolicyProvider == null;
 
-            if (_mvcOptions == null) 
+            if (!context.IsEffectivePolicy(this))
```
Comment (dougbu) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs]: Not sure why this check was behind the `AllowCombiningAuthorizeFilters` option. That required some seemingly-unrelated test changes.

In any case, should this be above the `ComputePolicyAsync()` call or in `OnAuthorizationAsync(...)`, prior to calling `GetEffectivePolicyAsync(...)`. The extra `async` state machines for this sync call seem wasteful.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.Core.Test/Authorization/AuthorizeFilterTest.cs [hunk 3]
```diff
@@ -31,13 +31,15 @@ public async Task DefaultConstructor_DeniesAnonymousUsers()
...
             var authorizationContext = GetAuthorizationContext(anonymous: true);
+
             // The type 'AuthorizeFilter' is both a filter by itself and also a filter factory.
             // The default filter provider first checks if a type is a filter factory and creates an instance of
             // this filter.
             var authorizeFilterFactory = new AuthorizeFilter();
             var filterFactory = authorizeFilterFactory as IFilterFactory;
             var authorizeFilter = (AuthorizeFilter)filterFactory.CreateInstance(
                 authorizationContext.HttpContext.RequestServices);
+            authorizationContext.Filters.Add(authorizeFilter);
```
Comment (dougbu) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.Core.Test/Authorization/AuthorizeFilterTest.cs]: Now required because the filter checks whether it's the effective one.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 4]
```diff
@@ -1718,11 +1719,12 @@ public async Task FromBody_JToken_ExcludedFromValidation()
...
 
             var testContext = ModelBindingTestHelper.GetTestContext(
-                request =>
+                updateRequest: request =>
                 {
                     request.Body = new MemoryStream(Encoding.UTF8.GetBytes("{ message: \"Hello\" }"));
                     request.ContentType = "application/json";
-                });
+                },
+                mvcOptions: options);
```
Comment (dougbu) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: Need to use the same `MvcOptions` instance was an odd fallout of removing `AllowValidatingTopLevelNodes`. The `ParameterBinderExtensions` helpers create `ModelMetadata` with a provider that uses these options. The default `MvcOptions` that `GetTestContext(...)` returns does not configure any binder providers.

**Comment by dougbu** [comment]
:up::date: just to rebase
@pranavkm & @rynowak please have a look. This PR is the only thing that's holding me up opening up one for #4866 part 3 of 3.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs (hunk 2 above)
Comment (pranavkm) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs]: Yeah, that seems wonky. There's an issue tracking cleaning this type up in 3.0: https://github.com/aspnet/AspNetCore/issues/4862

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs [hunk 5]
```diff
@@ -45,11 +29,8 @@ public void Configure(ApiBehaviorOptions options)
             ConfigureClientErrorMapping(options);
         }
 
-        public override void PostConfigure(string name, ApiBehaviorOptions options)
+        public void PostConfigure(string name, ApiBehaviorOptions options)
         {
-            // Let compatibility switches do their thing.
```
Comment (pranavkm) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs]: In 3.0, `InvalidModelStateResponseFactory` would default to returning a `ProblemDetails` instance (i.e. `options.InvalidModelStateResponseFactory = ProblemDetailsFactory;`) and `SuppressMapClientErrors` would have no bearing on it. Any user configured factory would take precedence over the default

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs [hunk 6]
```diff
@@ -53,7 +52,7 @@ public class ParameterBinder
         /// <param name="modelMetadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
         /// <param name="modelBinderFactory">The <see cref="IModelBinderFactory"/>.</param>
         /// <param name="validator">The <see cref="IObjectModelValidator"/>.</param>
-        /// <param name="mvcOptions">The <see cref="MvcOptions"/> accessor.</param>
+        /// <param name="mvcOptions">The <see cref="MvcOptions"/> accessor. This parameter is currently unused.</param>
```
Comment (pranavkm) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs]: I'd prefer leaving this doc comment as is. It'd be easy to miss updating this if we do start using it.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs [hunk 7]
```diff
@@ -39,10 +39,12 @@ internal class DefaultObjectValidator : ObjectModelValidator
...
                 validatorCache,
                 metadataProvider,
-                validationState);
-
-            visitor.MaxValidationDepth = _mvcOptions.MaxValidationDepth;
-            visitor.AllowShortCircuitingValidationWhenNoValidatorsArePresent = _mvcOptions.AllowShortCircuitingValidationWhenNoValidatorsArePresent;
+                validationState)
+            {
+                MaxValidationDepth = _mvcOptions.MaxValidationDepth,
+                AllowShortCircuitingValidationWhenNoValidatorsArePresent =
```
Comment (pranavkm) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs]: Might have missed when this was discussed in the issue, but would we keep this? I thought we would just default to doing this in 3.0. The flag was there because we just didn't feel confident enough about enabling the behavior super late in 2.2

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs [hunk 8]
```diff
@@ -67,7 +67,7 @@ public async Task AuthorizationPoliciesCombine()
             var request = new HttpRequestMessage(HttpMethod.Get, "http://localhost/Administration/EitherCookie");
             request.Headers.Add("Cookie", cookie2);
 
-            // Act 2: Will succeed because, with AllowCombiningAuthorizeFilters true, [Authorize] allows either cookie.
+            // Act 2: Will succeed because {Authorize] allows either cookie.
```
Comment (pranavkm) [inline src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs]: `[Authorize]`

**Review by pranavkm — APPROVED** [review]

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs (hunk 5 above)
Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs]: @pranavkm I don't disagree. But, are you asking for a change here?

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs (hunk 6 above)
Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs]: It's not clear to me this parameter will be used in the future. But, I could move this to `<remarks>`. Either way, the comment is hard to miss.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs (hunk 7 above)
Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs]: The change here is just cleanup. `AllowShortCircuitingValidationWhenNoValidatorsArePresent` goes away in part 3 of 3.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs (hunk 8 above)
Comment (dougbu) [reply src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs]: Is there a typo here I can't see?

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs (hunk 2 above)
Comment (rynowak) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs]: 👍

Comment (rynowak) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs]: Yeah I would imagine so.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs [hunk 9]
```diff
@@ -197,7 +197,7 @@ public virtual bool CanCreateInstance(Type targetType)
         /// <param name="bindingContext">The <see cref="ModelBindingContext"/>.</param>
         /// <remarks>
         /// <para>
-        /// This method should be called only when <see cref="MvcOptions.AllowValidatingTopLevelNodes" /> is
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: Should we also remove the implementations that handle `AllowValidatingTopLevelNodes` and just mark it obsolete and say that it doesn't do anything?

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json [hunk 10]
```diff
@@ -72,5 +72,35 @@
     "TypeId": "public abstract class Microsoft.AspNetCore.Mvc.ModelBinding.ObjectModelValidator : Microsoft.AspNetCore.Mvc.ModelBinding.Validation.IObjectModelValidator",
     "MemberId": "public abstract Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor GetValidationVisitor(Microsoft.AspNetCore.Mvc.ActionContext actionContext, Microsoft.AspNetCore.Mvc.ModelBinding.Validation.IModelValidatorProvider validatorProvider, Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidatorCache validatorCache, Microsoft.AspNetCore.Mvc.ModelBinding.IModelMetadataProvider metadataProvider, Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationStateDictionary validationState)",
     "Kind": "Addition"
+  },
+  {
+    "TypeId": "public class Microsoft.AspNetCore.Mvc.MvcOptions : System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Mvc.Infrastructure.ICompatibilitySwitch>",
+    "MemberId": "public System.Boolean get_AllowBindingHeaderValuesToNonStringModelTypes()",
+    "Kind": "Removal"
```
Comment (rynowak) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json]: We should consider leaving these implementations here, **or** coming up with a better design. I don't really relish us adding and removing this (and churning supressions) each release. 

Adding/removing properties is semantically meaningful, but this isn't

**Review by rynowak — APPROVED** [review]

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs (hunk 2 above)
Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs]: @rynowak should I reorder things in this PR or just add a comment in #4862?

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs (hunk 9 above)
Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: @rynowak why make the additional breaking change?

Users can disable validation of top-level nodes in a piecemeal fashion:
- replace the `IModelBinderProvider`s
- replace the `IModelMetadataProvider` with something that doesn't extend `ModelMetadataProvider`, and / or
- replace the `IObjectModelValidator` with something that doesn't extend `ObjectModelValidator`

But, ignoring `AllowValidatingTopLevelNodes` in the model binders would mean top-level validation could only be partially disabled.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json (hunk 10 above)
Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json]: This is about removing properties (specifically their getters and setters). We decided in an offline email thread to
> Completely remove the properties we no longer want to support i.e. those enabling most of the legacy behaviours.

Are you suggesting reopening that conversation?

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs (hunk 9 above)
Comment (rynowak) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: We don't intend to allow you to disable validation for parameters. If you don't like it, don't put validation attributes on parameters. We did a compat switch for this in case we broke functionality - which we did 😆 

I'm not suggesting we break lots of APIs. I'm suggesting that where the property/arg appears, we ignore it and document that it is ignored.

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json (hunk 10 above)
Comment (rynowak) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json]: Maybe I'm confused. I thought you also un-implemented the interfaces.

Comment (rynowak) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/breakingchanges.netcore.json]: This is what I'm talking about: https://github.com/aspnet/AspNetCore/pull/4628/files#diff-d9489227febd545e09ee512b868b9b8dL16

**Comment by dougbu** [comment]
@rynowak a couple of questions to you are outstanding&hellip;

For `[Authorize]`:
> should I reorder things in this PR or just add a comment in #4862?

For the model binders: Should I ignore all `allowValidatingTopLevelNodes` parameters and `AllowValidatingTopLevelNodes` properties except the pair in `CollectionModelBinder`? That would be a single inconsistency instead of
> the only parameter and property pair that could be ignored are in ComplexTypeModelBinder and DictionaryModelBinder. The inconsistency would be a bit ugly.

**Comment by dougbu** [comment]
@pranavkm I couldn't understand a couple of your comments&hellip;

For `ApiBehaviorOptionsSetup`:
> In 3.0, `InvalidModelStateResponseFactory` would default to returning a `ProblemDetails` instance (i.e. `options.InvalidModelStateResponseFactory = ProblemDetailsFactory;`) and `SuppressMapClientErrors` would have no bearing on it. Any user configured factory would take precedence over the default

For `GlobalAuthorizationFilterTest`:
> `[Authorize]`

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs (hunk 8 above)
Comment (pranavkm) [reply src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs]: There's a brace `{` at the start instead of `[`, no?

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs (hunk 5 above)
Comment (pranavkm) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs]: Basically, set `options.InvalidModelStateResponseFactory = ProblemDetailsFactory;` in `Configure` and remove the `PostConfigure` since it's no longer conditional.

Comment (dougbu) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApiBehaviorOptionsSetup.cs]: The `InvalidModelStateResponseFactory` setting remains conditional on `SuppressMapClientErrors` and that's a regular option that may be set at any point. Better to leave reading it 'til `PostConfigure(...)`.

File: src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs (hunk 8 above)
Comment (dougbu) [reply src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/GlobalAuthorizationFilterTest.cs]: Got it

**Comment by dougbu** [comment]
:up::date: to address one of @pranavkm's suggestions and to rebase
